src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -284,90 +284,109 @@ static Solution UnifyLinkedDocumentContents(Solution oldSolution, Solution newSo
                     continue;
 
-                // Now do the same for all added documents in a project.
-                ProjectId? relatedProjectIdHint = null;
-                foreach (var addedDocument in projectChanges.GetAddedDocuments())
-                    (newSolution, relatedProjectIdHint) = UpdateAddedDocumentToExistingContentsInSolution(newSolution, addedDocument, relatedProjectIdHint);
-
-                // now, for any changed document, ensure we go and make all links to it have the same text/tree.
-                foreach (var changedDocumentId in projectChanges.GetChangedDocuments())
-                    newSolution = UpdateExistingDocumentsToChangedDocumentContents(newSolution, changedDocumentId, seenChangedDocuments);
+                // Now do the same for all added and changed documents in a project.
+                addedDocumentIds.AddRange(projectChanges.GetAddedDocuments());
+                changedDocumentIds.AddRange(projectChanges.GetChangedDocuments());
             }
 
+            newSolution = UpdateAddedDocumentToExistingContentsInSolution(newSolution, addedDocumentIds);
+
+            // now, for any changed document, ensure we go and make all links to it have the same text/tree.
+            newSolution = UpdateExistingDocumentsToChangedDocumentContents(newSolution, changedDocumentIds);
+
             return newSolution;
         }
 
-        static (Solution newSolution, ProjectId? relatedProjectId) UpdateAddedDocumentToExistingContentsInSolution(
-            Solution solution, DocumentId addedDocumentId, ProjectId? relatedProjectIdHint)
+        static Solution UpdateAddedDocumentToExistingContentsInSolution(
+            Solution solution, ArrayBuilder<DocumentId> addedDocumentIds)
         {
-            Contract.ThrowIfTrue(addedDocumentId.ProjectId == relatedProjectIdHint);
+            ProjectId? relatedProjectIdHint = null;
+            using var _ = ArrayBuilder<(DocumentId, DocumentState)>.GetInstance(out var relatedDocumentIdsAndStates);
 
-            // Look for a related document we can create our contents from.  We only have to look for a single related
-            // doc as we'll be done once we update our contents to theirs.  Note: GetFirstRelatedDocumentId will also
-            // not search the project that addedDocumentId came from.  So this will help ensure we don't repeatedly add
-            // documents to a project, then look for related docs *within that project*, forcing the file-path map in it
-            // to be recreated for each document.
-            var relatedDocumentId = solution.GetFirstRelatedDocumentId(addedDocumentId, relatedProjectIdHint);
+            foreach (var addedDocumentId in addedDocumentIds)
+            {
+                // Ensure we don't search in addedDocumentId's project for the related document
+                if (addedDocumentId.ProjectId == relatedProjectIdHint)
+                    relatedProjectIdHint = null;
```

Comment:
is this change necessary?  (i'm guessing yes.  does GetFirstRelatedDocumentId basically assert to make sure the caller isn't doing something dumb?)

Reply:
Exactly. First line in SolutionState.GetFirstRelatedDocumentId:

Contract.ThrowIfTrue(documentId.ProjectId == relatedProjectIdHint);